File: js/allConstruct.js
// allConstruct(target, wordBank) return 2D array 
// containing all possible combinations
//
// allConstruct(purple, [purp, p, le, ur, purpl])
// [
// 		[purp, le],
// 		[p, ur, p, le]
// ]


function allConstruct(target, wordBank, memo = {}) {
	if (target in memo) return memo[target]
	if (target === '') return [[]]

	const result = []

	for (let word of wordBank) {
		const size = word.length
		if (target.slice(0, size) === word) {
			const suffixWays = allConstruct(target.slice(size), wordBank, memo)
			const targetWays = suffixWays.map(way => [word, ...way])
			result.push(...targetWays)
		}
	}

	memo[target] = result
	return result
}


console.log(allConstruct('skateboard', ['purp', 'p', 'le', 'ur', 'purpl']))
console.log(allConstruct('purple', ['purp', 'p', 'le', 'ur', 'purpl']))
console.log(allConstruct('abc', ['a', 'bc', 'ab', 'c']))

File: cpp/allConstruct.cpp
// allConstruct(target, wordBank) return 2D array
// containing all possible combinations 
//  
// allConstruct(purple, [purp, p, le, ur, purpl]) 
// [    
//      [purp, le],    
//      [p, ur, p, le]    
// ]    

#include <iostream>
#include <functional>
#include <algorithm>
#include <vector>
#include <map>

using namespace std;

vector<vector<string>> allConstruct(string target, const vector<string>& wordBank) {

	function< vector<vector<string>>(string, const vector<string>&) > impl;
	impl = [memo= map<string, vector<vector<string>>>() , &impl] (string target, const vector<string>& wordBank) mutable {
		if (auto ret_it = memo.find(target); ret_it != memo.end()) return ret_it->second;
		if (target == "") return vector<vector<string>>(1);

		vector<vector<string>> possibleCombo;
		for (auto word : wordBank) {
			if (target.compare(0, word.size(), word) == 0) {
				string suffix = target.substr(word.size(), target.size() - word.size());
				auto suffixCombo = impl(suffix, wordBank);
				for (auto& suffixWay: suffixCombo) {
					suffixWay.insert(suffixWay.begin(), word);
					// suffixWay.push_back(word);
					possibleCombo.push_back(move(suffixWay));
				}
			}
		}

		auto it = memo.emplace(target, move(possibleCombo));
		return it.first->second;
	};

	return impl(target, wordBank);
}


vector<vector<string>> allConstruct(string target, const vector<string>&& wordBank) {
	return allConstruct(target, wordBank);
}

int main() {
	// vector<string> wordBank {"p", "purp", "ur", "le", "purpl"};
	// auto result = allConstruct("purple", wordBank);
	auto result = allConstruct("purple", {"p", "purp", "ur", "le", "purpl"});

	for (auto& i : result) {
		// reverse(i.begin(), i.end());
		for (auto m : i)
			cout << m << " ";
		cout << endl;
	}
}
